### PR TITLE
Null checks for optional parameters 

### PR DESCRIPTION
**Pull request by tgsstdio**
Should be able to pass into NativeMethods calls AllocationCallbacks allocator as null

Out parameters should NOT be affected

### CODE REVIEW

File: tools/Generator/Generator.cs [hunk 1]
```diff
@@ -728,7 +732,14 @@ struct FixedParamInfo
 					bool isFixed = fixedParams != null &&  fixedParams.ContainsKey (name);
 					string paramName = isFixed ? "ptr" + name : name;
 
-					Write ("{0}{1}{2}", (isPointer && !isStruct && !isFixed) ? "&" : "", keywords.Contains (paramName) ? "@" + paramName : paramName, (!isFixed && (isStruct || isHandle)) ? ".m" : "");
+					if (isOptionalParam && isPointer && !isOut)
+					{
+						Write ("{0} != null ? {0}{1} : null", keywords.Contains (paramName) ? "@" + paramName : paramName, (!isFixed && (isStruct || isHandle)) ? ".m" : "");
```
Comment (jonpryor) [inline tools/Generator/Generator.cs]: Please split out the expression `!isFixed && (isStruct || isHandle)` into a separate variable and use the variable here and below.

Comment (jonpryor) [inline tools/Generator/Generator.cs]: The expression `keywords.Contains (paramName) ? "@" + paramName : paramName` should be split out into a separate helper method.

**Comment by tgsstdio** [comment]
Updated Generator.cs after to code review

**Comment by radekdoulik** [comment]
Thanks for the patch! I squashed the commits together and cherry picked.

File: tools/Generator/Generator.cs [hunk 2]
```diff
@@ -727,14 +731,28 @@ struct FixedParamInfo
...
+					{
+						Write ("{0}{1}{2}", (isPointer && !isStruct && !isFixed) ? "&" : "", GetSafeParameterName(paramName), useHandlePtr ? ".m" : "");
+					}
 				} else
 					Write ("{0}{1} {2}", isOut ? "out " : "", csType, keywords.Contains (name) ? "@" + name : name);
 			}
 
 			return outParams;
 		}
+
```
Comment (radekdoulik) [inline tools/Generator/Generator.cs]: Please don't add white space to empty lines

File: tools/Generator/Generator.cs [hunk 3]
```diff
@@ -691,6 +691,10 @@ struct FixedParamInfo
 				string name = param.Element ("name").Value;
 				string csType = GetTypeCsName (type);
 
+				var optional = param.Attribute ("optional");
+
+				bool isOptionalParam = (optional != null && optional.Value == "true");
+
```
Comment (radekdoulik) [inline tools/Generator/Generator.cs]: Please use the same coding style as surrounding code.

File: tools/Generator/Generator.cs [hunk 4]
```diff
@@ -727,14 +731,28 @@ struct FixedParamInfo
...
-					Write ("{0}{1}{2}", (isPointer && !isStruct && !isFixed) ? "&" : "", keywords.Contains (paramName) ? "@" + paramName : paramName, (!isFixed && (isStruct || isHandle)) ? ".m" : "");
+					if (isOptionalParam && isPointer && !isOut)
+					{
+						Write ("{0} != null ? {0}{1} : null", GetSafeParameterName(paramName), useHandlePtr ? ".m" : "");
+					}
+					else
+					{
+						Write ("{0}{1}{2}", (isPointer && !isStruct && !isFixed) ? "&" : "", GetSafeParameterName(paramName), useHandlePtr ? ".m" : "");
+					}
 				} else
```
Comment (radekdoulik) [inline tools/Generator/Generator.cs]: As above

**Pull request closed without merging**